### PR TITLE
Support multiclass classification

### PR DESCRIPTION
**Pull request by wxchan**
to solve #43 

I'd like to open this PR to let more people see and help build or test on this feature. [Lastest version](https://github.com/wxchan/LightGBM/tree/dev).

You need to set `objective = multiclass`, `metric = multi_logloss(or multi_error)` and `num_class =` in train.conf, `num_class =` in predict.conf is not needed any more. There is an example in example/multiclass_classification.

### CODE REVIEW

**Comment by guolinke** [comment]
@wxchan Thanks for your nice work.
However, this seems will have conflict with my PR for the prepare of c_api. I refine some interface on boosting.h and metric.h. 
Maybe you can continue after I finish the changes for interface refine.
Thanks again.

**Comment by chivee** [comment]
Hi, @wxchan , you can rebase master for the new c_api.

BTW, could you please also give some test example for this PR?

**Comment by wxchan** [comment]
@guolinke I fixed the conflicts according to your new commits.

@chivee I already added an example in examples/multiclass_classification. You can run it and use compare.py [here](https://github.com/wxchan/LightGBM/commit/3449b0c71ae3cf780383a7ae513b75624e7463b2) to compare its performance with xgboost.

**Comment by wxchan** [comment]
A small bug when save the model. I think [this line](https://github.com/Microsoft/LightGBM/blob/master/src/application/application.cpp#L251) is not necessary. Because you already pop_back the trees if meets early stopping. The remaining trees will save at next SaveModelToFile(). After my commit, some trees will save twice, can't figure out why.

Btw, is it necessary to delete back() before pop_back()? I thought pop_back() ([see docs](http://www.cplusplus.com/reference/vector/vector/pop_back/)) will delete it anyway.

**Comment by guolinke** [comment]
@wxchan I think pop_back() cannot free the resource of a pointer. 
refer to: http://stackoverflow.com/questions/10696124/vectorclear-pop-back-wont-run-destructor
https://www.daniweb.com/programming/software-development/threads/170492/does-std-vector-s-pop-back-free-the-memory
http://stackoverflow.com/questions/10696124/vectorclear-pop-back-wont-run-destructor

**Comment by guolinke** [comment]
@wxchan  to avoid save repeated model, you need to change [this](https://github.com/wxchan/LightGBM/blob/dev/src/boosting/gbdt.cpp#L328-L333) as well.

**Comment by wxchan** [comment]
@guolinke Yes, I know, but I can't figure out how. (I already changed ` early_stopping_round_` to `early_stopping_round_ * num_class_`, but it's not where the problem is)

Actually, I am not sure your origin code is right. For example, set early_stopping_round_ = 10, and assume when iter = 50, it meets early stopping criteria. In this situation, at iter 50, you already write first 40 trees to file. Then because of you pop last 10 trees in TrainOneIter(), the models_.size() = 40. So, rest = 40 - 10 = 30. Then at line 336, saved_model_size_ = rest = 30. So, in line 341-348, 31-40 trees are saved again. 

You can try to remove the SaveModelToFile() line I mentioned above and see it works. At least the result seems right.

**Comment by guolinke** [comment]
@wxchan Thanks, you are right. I will fix this directly, then merge your PR after some other reviews.

**Comment by guolinke** [comment]
@wxchan I fix this by change this https://github.com/Microsoft/LightGBM/blob/master/src/boosting/gbdt.cpp#L321 .
Use "iter_" is not so elegant.

File: src/metric/multiclass_metric.hpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  explicit MulticlassMetric(const MetricConfig& config) {
+      num_class_ = config.num_class;
+  }
+
+  virtual ~MulticlassMetric() {
+
+  }
+
+  void Init(const char* test_name, const Metadata& metadata, data_size_t num_data) override {
+    name_ = test_name;
```
Comment (guolinke) [inline src/metric/multiclass_metric.hpp]: name_ = test_name+metric_name..

File: src/metric/multiclass_metric.hpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        }
+        // add loss
+        sum_loss += PointWiseLossCalculator::LossOnPoint(label_[i], rec);
+      }
+    } else {
+      #pragma omp parallel for schedule(static) reduction(+:sum_loss)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        std::vector<score_t> rec(num_class_);
+        for (int k = 0; k < num_class_; ++k) {
+          rec[k] = score[k * num_data_ + i];
```
Comment (guolinke) [inline src/metric/multiclass_metric.hpp]: don't need softmax before calculate logloss?

File: src/objective/multiclass_objective.hpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  void GetGradients(const score_t* score, score_t* gradients, score_t* hessians) const override {
+    if (weights_ == nullptr) {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        std::vector<score_t> rec(num_class_);
+        int ilabel = static_cast<int>(label_[i]);  
+        if (ilabel < 0 || ilabel >= num_class_) {
+            Log::Fatal("Label must be in [0, num_class)");
```
Comment (guolinke) [inline src/objective/multiclass_objective.hpp]: This checking should be done in "Init".

File: src/objective/multiclass_objective.hpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    label_ = metadata.label();
+    weights_ = metadata.weights();
+  }
+
+  void GetGradients(const score_t* score, score_t* gradients, score_t* hessians) const override {
+    if (weights_ == nullptr) {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        std::vector<score_t> rec(num_class_);
+        int ilabel = static_cast<int>(label_[i]);
```
Comment (guolinke) [inline src/objective/multiclass_objective.hpp]: will cast to int every time. Maybe using an int array to cache this is better

File: src/objective/multiclass_objective.hpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            gradients[k * num_data_ + i] = p * weights_[i];
+          }
+          hessians[k * num_data_ + i] = 2.0f * p * (1.0f - p) * weights_[i];
+        }
+      }
+    }
+  }
+
+  float GetSigmoid() const override {
+    return -1.0;
```
Comment (guolinke) [inline src/objective/multiclass_objective.hpp]: -1.0f

File: src/objective/multiclass_objective.hpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  int num_class_;
+  /*! \brief Pointer of label */
+  const float* label_;
+  /*! \brief True if using unbalance training */
+  bool is_unbalance_;
+  /*! \brief Values for multiclass labels */
+  int* label_val_;
+  /*! \brief Weights for multiclass labels */
+  score_t* label_weights_;
+  /*! \brief Weights for data */
```
Comment (guolinke) [inline src/objective/multiclass_objective.hpp]: remove unused variables. is_unbalance_, label_val_, label_weights_

File: src/boosting/score_updater.hpp [hunk 7]
```diff
@@ -18,12 +18,12 @@ class ScoreUpdater {
   * \brief Constructor, will pass a const pointer of dataset
   * \param data This class will bind with this data set
   */
-  explicit ScoreUpdater(const Dataset* data)
-    :data_(data) {
+  explicit ScoreUpdater(const Dataset* data, int num_class)
+    :data_(data), num_class_(num_class) {
```
Comment (guolinke) [inline src/boosting/score_updater.hpp]: Should we really need to pass "num_class" to score_updater?
I think we can add an parameter "offset" in every "AddScore". 
And pass "num_data_*num_class_" as size in constructor

File: src/boosting/gbdt.h [hunk 8]
```diff
@@ -180,6 +199,10 @@ class GBDT: public Boosting {
   data_size_t bag_data_cnt_;
   /*! \brief Number of traning data */
   data_size_t num_data_;
+  /*! \brief Number of classes */
+  int num_class_;
+  /*! \brief Current class */
+  int curr_class_;
```
Comment (guolinke) [inline src/boosting/gbdt.h]: really use this?  Or add "curr_class" as parameter in these functions who use "curr_class"?

File: src/boosting/gbdt.cpp [hunk 9]
```diff
@@ -471,4 +496,4 @@ std::vector<int> GBDT::PredictLeafIndex(const float* value, int num_used_model)
   return ret;
 }
 
-}  // namespace LightGBM
+}  // namespace LightGBM
```
Comment (guolinke) [inline src/boosting/gbdt.cpp]: an new line ...

File: include/LightGBM/utils/common.h [hunk 10]
```diff
@@ -336,6 +336,22 @@ static inline int64_t Pow2RoundUp(int64_t x) {
   return 0;
 }
 
+inline void Softmax(std::vector<float>* p_rec) {
```
Comment (xuehui1991) [inline include/LightGBM/utils/common.h]: I wonder what is the abbreviation of "p_rec"? can you rename this parameter?

File: include/LightGBM/utils/common.h [hunk 11]
```diff
@@ -336,6 +336,22 @@ static inline int64_t Pow2RoundUp(int64_t x) {
...
+  std::vector<float> &rec = *p_rec;
+  float wmax = rec[0];
```
Comment (xuehui1991) [inline include/LightGBM/utils/common.h]: what's meaning of "wmax"?

File: src/application/predictor.hpp [hunk 12]
```diff
@@ -120,7 +133,20 @@ class Predictor {
     };
 
     std::function<std::string(const std::vector<std::pair<int, float>>&)> predict_fun;
-    if (is_predict_leaf_index_) {
+    if (num_class_ > 1) {
+      predict_fun = [this](const std::vector<std::pair<int, float>>& features){
+        std::vector<float> prediction = PredictMulticlassOneLine(features);
+        std::stringstream result_ss;
```
Comment (xuehui1991) [inline src/application/predictor.hpp]: rename to "result_stream_buf"?

File: src/boosting/gbdt.cpp [hunk 13]
```diff
@@ -146,52 +152,60 @@ void GBDT::Bagging(int iter) {
...
-  if (new_tree->num_leaves() <= 1) {
-    Log::Info("Can't training anymore, there isn't any leaf meets split requirements.");
-    return true;
-  }
-  // shrinkage by learning rate
-  new_tree->Shrinkage(gbdt_config_->learning_rate);
-  // update score
-  UpdateScore(new_tree);
-  UpdateScoreOutOfBag(new_tree);
+
```
Comment (xuehui1991) [inline src/boosting/gbdt.cpp]: why add a blank line here?

File: src/boosting/gbdt.h (hunk 8 above)
Comment (wxchan) [inline src/boosting/gbdt.h]: curr_class_ is used in Bagging(), UpdateScoreOutOfBag(), TrainOneIter() and UpdateScore(). It indicates current tree_learner we are using. It works just like iter_.

File: include/LightGBM/utils/common.h (hunk 11 above)
Comment (wxchan) [inline include/LightGBM/utils/common.h]: @xuehui1991 to be honest, I don't know. I copied it from [xgboost](https://github.com/dmlc/xgboost/blob/master/src/common/math.h#L30). The result is same with or without it.

File: src/io/config.cpp [hunk 14]
```diff
@@ -45,7 +45,27 @@ void OverallConfig::Set(const std::unordered_map<std::string, std::string>& para
...
+          Log::Fatal("You should specify number of class(>=2) for multiclass training.");
+      }
+  }
+  else {
+      if (task_type == TaskType::kTrain && num_class_test != 1){
+          Log::Fatal("Number of class must be 1 for non-multiclass training.");
+      }      
+  }
+  for (std::string metric_type : metric_types){
+        bool metric_type_multiclass = ( metric_type == "multi_logloss" || metric_type == "multi_error");
```
Comment (guolinke) [inline src/io/config.cpp]: It is better to use  metric_type == std::string("multi_logloss" )

File: src/io/config.cpp [hunk 15]
```diff
@@ -45,7 +45,27 @@ void OverallConfig::Set(const std::unordered_map<std::string, std::string>& para
   if (boosting_type == BoostingType::kGBDT) {
     boosting_config = new GBDTConfig();
   }
-
+  
+  bool objective_type_multiclass = (objective_type == "multiclass");
+  int num_class_test = 1;
```
Comment (chivee) [inline src/io/config.cpp]: please using assert for runtime logic check

File: src/metric/multiclass_metric.hpp [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+        }
+        // add loss
+        sum_loss += PointWiseLossCalculator::LossOnPoint(label_[i], rec) * weights_[i];
+      }
+    }
+    score_t loss = sum_loss / sum_weights_;
+    return std::vector<score_t>(1, loss);;
```
Comment (xuehui1991) [inline src/metric/multiclass_metric.hpp]: why there are two "；"？

File: src/boosting/gbdt.h (hunk 8 above)
Comment (guolinke) [inline src/boosting/gbdt.h]: I know this. My point is should we really need to use "curr_class_" as a member variable?
why not pass "curr_class" as the argument of these functions?

File: src/io/config.cpp (hunk 15 above)
Comment (guolinke) [inline src/io/config.cpp]: @wxchan you can direct use "GetInt(params, "num_class", &num_class);" then test the conflict in the function "CheckParamConflict"

File: src/objective/multiclass_objective.hpp [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+#define LIGHTGBM_OBJECTIVE_MULTICLASS_OBJECTIVE_HPP_
+
+#include <LightGBM/objective_function.h>
+
+#include <cstring>
+#include <cmath>
+
+namespace LightGBM {
+/*!
+* \brief Objective funtion for multiclass classification
```
Comment (xuehui1991) [inline src/objective/multiclass_objective.hpp]: function?

File: src/objective/multiclass_objective.hpp [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+    num_class_ = config.num_class;
+  }
+  
+  ~MulticlassLogloss() {}
+  
+  void Init(const Metadata& metadata, data_size_t num_data) override {
+    num_data_ = num_data;
+    label_ = metadata.label();
+    weights_ = metadata.weights();
+    ilabel_ = new int[num_data_];
```
Comment (xuehui1991) [inline src/objective/multiclass_objective.hpp]: can you change the name "ilabel_"?

File: src/application/application.cpp [hunk 19]
```diff
@@ -269,6 +269,7 @@ void Application::Predict() {
 void Application::InitPredict() {
   boosting_ =
     Boosting::CreateBoosting(config_.io_config.input_model.c_str());
+  boosting_ -> SetNumberOfClass(config_.boosting_config->num_class);
```
Comment (wxchan) [inline src/application/application.cpp]: @guolinke it's a great idea. We don't need num_class in predict.conf then.

File: src/application/predictor.hpp [hunk 20]
```diff
@@ -33,6 +33,7 @@ class Predictor {
       num_used_model_(num_used_model) {
     boosting_ = boosting;
     num_features_ = boosting_->MaxFeatureIdx() + 1;
+    num_class_ = boosting_->NumberOfClass();
```
Comment (guolinke) [inline src/application/predictor.hpp]: "num_class_" is not used in predictor?

File: src/objective/multiclass_objective.hpp [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+        }  
+        Common::Softmax(&rec);
+        for (int k = 0; k < num_class_; ++k) {
+          float p = rec[k];
+          if (label_int_[i] == k) {
+            gradients[k * num_data_ + i] = (p - 1.0f) * weights_[i];
+          } else {
+            gradients[k * num_data_ + i] = p * weights_[i];
+          }
+          hessians[k * num_data_ + i] = 2.0f * p * (1.0f - p) * weights_[i];
```
Comment (guolinke) [inline src/objective/multiclass_objective.hpp]: @wxchan Did you know why here need to multiply ```2.0``` ?  From my calculation, it didn't has this factor.

Comment (wxchan) [inline src/objective/multiclass_objective.hpp]: I referred it to xgboost a little: https://github.com/dmlc/xgboost/blob/master/src/objective/multiclass_obj.cc#L66

I am not sure why it needs to multiply 2 now, seem necessary.

Comment (Laurae2) [inline src/objective/multiclass_objective.hpp]: In Newton method, the second derivative for it is: `p*(1-p)` but it is only for the binary case.

For softmax it seems it comes from another approximation method, on this paper they use 2 on the `B^-1` part for the upper bounding computation so the formula can be used as the hessian instead: http://www.ism.ac.jp/editsec/aism/pdf/044_1_0197.pdf (formula c and d, last page)

That `2` does not show up in the binary case (Newton method).

Comment (guolinke) [inline src/objective/multiclass_objective.hpp]: Thanks!

Comment (guolinke) [inline src/objective/multiclass_objective.hpp]: I found this from sklearn.gbm: https://github.com/scikit-learn/scikit-learn/blob/919b4a8fbdf0e313fb702bd083abb31d67f7d8f9/sklearn/ensemble/gradient_boosting.py#L566-L584 
It seems it multiplies the `(num_class-1)/num_class` .

Comment (Laurae2) [inline src/objective/multiclass_objective.hpp]: In matlab it seems `1/2` for softmax: https://fr.mathworks.com/matlabcentral/fileexchange/55863-logistic-regression-for-classification?focused=6032429&tab=function

Newton Raphson / IRLS: `f(x) = f(x_t) + df(x - x_t)/dx_t + (1/2)*(x - x_t)^T * d²f(x_t * (x - x_t))/(dxdxT) + o(||x - x_t||²)`.

As long as the number is not larger than the maximum allowed (`2` apparently on the paper I linked before) it should converge.
